### PR TITLE
Whisk

### PR DESCRIPTION
**Pull request by dapplion**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-xxxx.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+discussions-to: TBD
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-07-16
+---
+
+## Abstract
+
+Upgrades block proposer election mechanism Whisk, a single secret leader election (SSLE) protocol. Currently block proposers are publicly known in advance, sufficiently to allow sequential DoS attacks that could disable Ethereum. This upgrade makes the next block proposer unknown until its block is published.
```
Comment (asn-d6) [inline EIPS/eip-xxxx.md]:
```suggestion
Upgrades block proposer election mechanism to Whisk, a single secret leader election (SSLE) protocol. Currently block proposers are publicly known sufficiently in advance to allow sequential DoS attacks that could disable Ethereum. This upgrade makes the next block proposer unknown until its block is published.
```

File: EIPS/eip-xxxx.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+This requires no changes to the Execution Layer.
+
+### Consensus layer
+
+The protocol can be summarized in the following concurrent steps:
+
+- Each validator registers a tracker on their first proposal after the fork
+- At the start of a shuffling phase a list of candidate trackers is selected using public randomness from RANDAO
+- During one shuffling phase each proposer shuffles a subset of the candidate trackers using private randomness
```
Comment (asn-d6) [inline EIPS/eip-xxxx.md]:
```suggestion
- During the shuffling phase each proposer shuffles a subset of the candidate trackers using private randomness
```

File: EIPS/eip-xxxx.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+- After one shuffling phase an ordered list of proposer trackers is selected from the candidate set using RANDAO
```
Comment (asn-d6) [inline EIPS/eip-xxxx.md]:
```suggestion
- After the shuffling phase an ordered list of proposer trackers is selected from the candidate set using RANDAO
```

File: EIPS/eip-xxxx.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+Specification of the proposed change can be found in [`/_features/whisk/beacon-chain.md`](https://github.com/ethereum/consensus-specs/blob/a39abe388bc2d1abd5b4fd62fd18aed497956b30/specs/_features/whisk/beacon-chain.md).. In summary:
+
+- Update `BeaconState` with fields needed track validator trackers, commitments and the two rounds of candidates election.
```
Comment (asn-d6) [inline EIPS/eip-xxxx.md]:
```suggestion
- Update `BeaconState` with fields needed to track validator trackers, commitments and the two rounds of candidates election.
```

File: EIPS/eip-xxxx.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+- Add `select_whisk_candidate_trackers` to compute the next vector of candidates from the validator set.
+- Add `select_whisk_proposer_trackers` to compute the next vector of proposers from current candidates.
+- Add `process_whisk_updates` to epoch processing logic.
```
Comment (asn-d6) [inline EIPS/eip-xxxx.md]:
```suggestion
- Add `process_whisk_updates` to the epoch processing logic.
```

File: EIPS/eip-xxxx.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+- Add `process_whisk_opening_proof` to validate block proposer has knowledge of this slot's elected tracker.
```
Comment (asn-d6) [inline EIPS/eip-xxxx.md]:
```suggestion
- Add `process_whisk_opening_proof` to check that the block proposer has knowledge of this slot's elected tracker.
```

File: EIPS/eip-xxxx.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+## Rationale
+
+### Fields per validator
+
+Whisk requires to append one tracker `(rG,krG)` and one commitment `kG` per validator. Both are updated at max once through a validator's lifetime.
+
+Trackers are registered with a randomized base `(rG,krG)` to make it harder for adversaries to track them through shuffling gates. It can become an issue if the set of honest shufflers is small.
+
+Each tracker must be binded to a validator's identity to prevent multiple parties to claim the same proposer slot. Otherwise it would allow proposers to sell their proposer slot, and cause fork-choice issues if two competing blocks appear.
```
Comment (asn-d6) [inline EIPS/eip-xxxx.md]: Maybe this is handled by the section below?

File: EIPS/eip-xxxx.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+Whisk does identity binding by storing a commitment to the tracker's secret `kG` in the validator record. Storing the commitment also allows to ensures the uniqueness of `k`.
+
+Identity binding can be achieved by forcing the hash prefix of `hash(kG)` to match its validator index. However, validators would have to brute force `k` making bootstrap of the system harder to participants with less computational resources.
+
+Identity binding can also be achieved by setting `k = hash(nonce + pubkey)`. However, proposers will need to reveal `k` and be de-anonimized for repeated proposals on adjacent shuffling phases.
+
+### Alternative: non-single secret election
+
+Secret non-single leader election is based on protocol engineering rather than cryptography, thus much simpler and cheap than Whisk. However it complicates the fork-choice and opens it up to potential MEV time-buying attacks, making it an unsuitable option at the time of writing.
```
Comment (asn-d6) [inline EIPS/eip-xxxx.md]: Would be nice to link to the [ethresearch post](https://ethresear.ch/t/secret-non-single-leader-election/11789) but IIRC you can't add links to EIPs?

File: EIPS/eip-xxxx.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+The shuffling strategy is analyzed in a companion paper and considered sufficiently safe for Whisk's usecase. Data and computational complexity of this EIP is significant but constant, thus does not open new DoS vectors. 
+
+### Anonymity set
+
+The anonymity set in Whisk equals `WHISK_CANDIDATE_TRACKERS_COUNT - WHISK_PROPOSER_TRACKERS_COUNT` (8,192) which are the candidates that did not get selected as proposers. That count of validators correspond to a smaller number of p2p nodes. Assuming a Pareto principle where "20% of the nodes run 80% of the validators" the anonymity corresponds to 2,108 nodes on average. A bigger candidate pool could make the shuffling strategy unsafe, while shuffling more trackers per round would increase the cost of the ZK proofs.
+
+### RANDAO biasing
+
+Whisk uses RANDAO in the candidate selection and proposer selection events, and is susceptible to potential RANDAO biasing attacks by malicious proposers. Whisk security could be made identical to the status quo by spreading the selection events over and entire shuffling period. However, status quo security is not ideal either and it would complicate the protocol further.
```
Comment (asn-d6) [inline EIPS/eip-xxxx.md]:
```suggestion
Whisk uses RANDAO in the candidate selection and proposer selection events, and is susceptible to potential RANDAO biasing attacks by malicious proposers. Whisk security could be made identical to the status quo by spreading the selection events over the entire shuffling period. However, doing so would complicate the Whisk protocol further, and status quo security is not ideal either.
```

**Review by asn-d6 — COMMENTED** [review]
A first pass

File: EIPS/eip-xxxx.md (hunk 8 above)
Comment (dapplion) [reply EIPS/eip-xxxx.md]: That would be great, but not allowed with current EIP rules

**Comment by dapplion** [comment]
Closing for https://github.com/ethereum/EIPs/pull/7441

**Pull request closed without merging**